cvc5_pythonic_api/cvc5_pythonic.py

```diff
@@ -1727,6 +1734,793 @@ def Or(*args):
+        >>> s.is_string()
+        False
+        """
+        return isinstance(self, StringSortRef)
+
+
+class SeqRef(ExprRef):
+    """Sequence Expressions"""
+
+    def sort(self):
+        return _sort(self.ctx, self.ast)
+
+    def isSequence(self):
+        return True
+
+    def __add__(self, other):
+        """Concatenation of two sequences.
+
+        >>> s,t = Consts('s t',SeqSort(RealSort()))
+        >>> s+t
+        Concat(s, t)
+        >>> (s+t).sort()
+        (Seq Real)
+        """
+        return Concat(self, other)
+
+    def __radd__(self, other):
+        return Concat(other, self)
+
+    def __getitem__(self, i):
+        """Return the SMT expression `self[arg]`
+
+        >>> seq = Const('seq',SeqSort(IntSort()))
+        >>> seq[0]
+        seq [] 0
+        >>> seq[0].sort()
+        Int
+        """
+        if _is_int(i):
+            i = IntVal(i, self.ctx)
+        return _to_expr_ref(
+            self.ctx.solver.mkTerm(Kind.SEQ_NTH, self.ast, i.ast), self.ctx
+        )
+
+    def at(self, i):
+        """Return the sequence at index i
+
+        >>> seq = Const('seq',SeqSort(IntSort()))
+        >>> seq.at(0)
+        At(seq, 0)
+        >>> seq.at(0).sort()
+        (Seq Int)
+        """
+        if _is_int(i):
+            i = IntVal(i, self.ctx)
+        return SeqRef(self.ctx.solver.mkTerm(Kind.SEQ_AT, self.ast, i.ast), self.ctx)
+
+    def is_string(self):
+        return isinstance(self, StringRef)
+
+    def as_string(self):
+        """Return a string representation of sequence expression.
+
+        >>> s = Unit(IntVal(1)) + Unit(IntVal(2))
+        >>> s.as_string()
+        '(str.++ (seq.unit 1) (seq.unit 2))'
+        """
+        return str(self.ast)
+
+    def is_string_value(self):
+        """Return True if self is a string value
+
+        >>> s = String('s')
+        >>> s.is_string_value()
+        False
+        >>> t = StringVal('t')
+        >>> t.is_string_value()
+        True
+        """
+        return self.is_string() and self.ast.getKind() == Kind.CONST_STRING
+
+
+class StringSortRef(SeqSortRef):
+    """String sort."""
+
+    def cast(self, val):
+        """Try to cast `val` as a String.
+
+        >>> string_sort = StringSort()
+        >>> string_sort.cast(10)
+        "10"()
+        >>> string_sort.cast('hello')
+        "hello"()
+        >>> string_sort.cast(10.5)
+        "10.5"()
+        """
+        if is_expr(val):
+            if debugging():
+                _assert(self.ctx == val.ctx, "Context mismatch")
+            val_s = val.sort()
+            if self.eq(val_s):
+                return val
+            if val_s.is_bool() and self.is_int():
+                return If(val, 1, 0)
+            if val_s.is_bool() and self.is_real():
+                return ToReal(If(val, 1, 0))
+            if val_s.is_int() and self.is_real():
+                return ToReal(val)
+            if debugging():
+                _assert(False, "SMT Integer/Real expression expected")
```

Comment:
This can't be right, because none of these give string terms.

@yoni206 : are there any kinds/operators that convert from non-string sorts to string sorts? Perhaps from integers to the length-1 string that has the character given by the integer?

In this method, we should use such conversion kinds/operators.

Reply:
Yeah I think that the message in the assertion is not up to date.
In particular, we can see that the conditions above it include other types (like bool).

And yes, the doctest includes an example of a conversion from numbers:
https://github.com/cvc5/cvc5_pythonic_api/blob/ebdf7eb3c8eb330309fddd5bb65f6b8104b9c569/cvc5_pythonic_api/cvc5_pythonic.py#L1845

And the function handles this:
https://github.com/cvc5/cvc5_pythonic_api/blob/ebdf7eb3c8eb330309fddd5bb65f6b8104b9c569/cvc5_pythonic_api/cvc5_pythonic.py#L1856

@LaorS maybe a solution is just to add "Bool"? That is, "Integer/Real/Bool"?

Reply:
To clarify my comment: I think this implementation is correct when `val` is a primitive python type (which is why the doc-tests work). However, in the case where `is_expr(val)` is true, I think this implementation is not right In particular, one would expect that `is_string(StringSort().cast(val))` would be true, but that isn't the case here.

To put it succinctly, I think this cast is correct for *concrete* inputs, but incorrect for *symbolic* ones.

Reply:
Yes, it's not working when `val` is an expression (unless it's a string expression).

For numbers there is a function that returns a string representation of it, I can probably use it. But I don't see anything similar for Booleans or other non-string sorts.

Reply:
> For numbers there is a function that returns a string representation of it, I can probably use it. But I don't see anything similar for Booleans or other non-string sorts.

Sounds good. Just numbers, then.